TimeOfFlight/src/TOFEstimators.cc

```diff
@@ -1,633 +1,227 @@
 
-    if (x.size() <= 1) return 0.;
+    PIDHandler pidHandler( pfos );
+    int algoID = pidHandler.addAlgorithm( name(), _outputParNames );
 
-    TGraphErrors gr(x.size(), &x[0], &y[0], &xErr[0], &yErr[0]);
-    gr.Fit("pol1", "Q");
 
-    return gr.GetFunction("pol1")->GetParameter(0);
-}
+    for (int i=0; i<pfos->getNumberOfElements(); ++i){
+        ReconstructedParticle* pfo = static_cast <ReconstructedParticle*> ( pfos->getElementAt(i) );
 
+        int nClusters = pfo->getClusters().size();
+        int nTracks = pfo->getTracks().size();
 
-double TOFEstimators::getTOFClosestFit(const Track* track, const Cluster* cluster){
-    const TrackState* ts = track->getTrackState(TrackState::AtCalorimeter);
-    dd4hep::rec::Vector3D trackAtCaloPos = ts->getReferencePoint();
-    dd4hep::rec::Vector3D trackAtCaloMom = getMomAtCalo(track);
-
-    std::map<int, double> dToImpact;
-    std::map<int, double> hitTime;
-    std::map<int, double> dToLineMin;
-    for(int i=0; i<_maxLayerNum; ++i) dToImpact[i] = 0.;
-    for(int i=0; i<_maxLayerNum; ++i) hitTime[i] = 0.;
-    for(int i=0; i<_maxLayerNum; ++i) dToLineMin[i] = std::numeric_limits<double>::max();
-
-    for (const auto& hit : cluster->getCalorimeterHits() ){
-        CHT hitType( hit->getType() );
-        bool isEcal = (hitType.caloID() == CHT::ecal);
-        int layer = hitType.layer();
-        if (!isEcal || layer >= _maxLayerNum) continue;
-
-        dd4hep::rec::Vector3D hitPos = hit->getPosition();
-        double dToLine = (hitPos - trackAtCaloPos).cross(trackAtCaloMom.unit()).r();
-        if( dToLine < dToLineMin[layer] ){
-            dToLineMin[layer] = dToLine;
-            dToImpact[layer] = (hitPos - trackAtCaloPos).r();
-            hitTime[layer] = hit->getTime();
+        if( nClusters != 1 || nTracks != 1){
+            // Analyze only simple pfos. Otherwise write dummy zeros
+            vector<float> results{0., 0., 0.};
+            pidHandler.setParticleID(pfo , 0, 0, 0., algoID, results);
+            continue;
         }
+        Track* track = pfo->getTracks()[0];
+        Cluster* cluster = pfo->getClusters()[0];
+
+        ///////////////////////////////////////////////////////////////
+        // This part calculates track length and momentum harmonic mean
+        ///////////////////////////////////////////////////////////////
+        vector<Track*> subTracks = getSubTracks(track);
+        vector<TrackStateImpl> trackStates = getTrackStatesPerHit(subTracks, _trkSystem, _extrapolateToEcal, _bField);
+
+        double trackLength = 0.;
+        double harmonicMom = 0.;
+        int nTrackStates = trackStates.size();
+        for( int j=1; j < nTrackStates; ++j ){
+            //for the last hit we check which track length formula to use
+            if (j == nTrackStates - 1){
+                double nTurns = getHelixNRevolutions( trackStates[j-1], trackStates[j] );
+                if( nTurns > 0.5 ){
```

Comment:
Here I assumed that TPC hits are very close to each other so they can't have `>0.5` curls between them.

It is hard to think of an example where it can be the case...

The only example case I can think of is when track goes in&out of the inner edge of the TPC. Then the distance between two TPC hits can be larger than usual. On the image bellow, you can see that neighboring TPC hits of the two subtracks are far away, because part of the track is not in the TPC.

But still it should be always limited to be `< 0.5` curls by the purely geometrical reasons...

<img src="https://user-images.githubusercontent.com/28848011/136155900-5006e26e-c50e-4c5d-a809-c639ab43ab69.png" width="600"/>

But I think checking this for every pair will not harm for the robustness, so I will change this to check for every pair of hits as you proposed.

Maybe this is not ideal for the timewise performance, but I think this is the irrelevant for the time being.

Reply:
I think, unless we discover that this is really a bottleneck doing the check for all track states makes the code more robust **and** easier to read. Thanks for changing this to do that.